### PR TITLE
Rayleigh Scattering Model

### PR DESCRIPTION
**Pull request by whokion**
@sethrj  I created a new MR for Rayleigh (a part of Issue #1).  Before starting a full review, I would like to get general comments.  One potential change may be that Rayleigh parameters (currently inside RayleighData) can be read from a file during run time.

### CODE REVIEW

**Comment by sethrj** [comment]
@whokion On first glance this looks pretty good! I think that for now we should only rely on external datasets where they already exist -- e.g. any of the $G4XXX datasets -- but for more directly derived data I think it's good to just include it in the code, with appropriate documentation of course.

I'll give a second look this evening for high-level comments.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighInteractor.i.hh:53]: For our existing datasets, rather than copying all elements to the GPU and indexing into them with `Z - 1`, we've looped over `MaterialParams::num_elements()` and copied over only the elements in use in the current problem, and indexed into collections with `element_id`.

In practice I don't think it'll make too much of a difference vs copying over everything, but it's definitely safer to access the data via `element_id` rather than an offset Z integer.

File: src/physics/em/detail/RayleighData.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighData.hh:21]: Can you also add a comment about what the units are?

**Review by sethrj — CHANGES_REQUESTED** [review]
Thanks @whokion , this looks very good for the initial PR  --  I've added a couple of high-level comments about the parameter data storage.

File: src/physics/em/detail/RayleighData.hh
Comment (whokion) [reply src/physics/em/detail/RayleighData.hh:27]: Moved the Rayleigh angular parameters from RayleighData.hh (deleted) to Rayleigh.hh and Rayleigh.cc (newly added).

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh]: Add MaterialParams in RayleighModel and store angular parameters only for need element with element_id.

File: src/physics/em/detail/RayleighData.hh
Comment (whokion) [reply src/physics/em/detail/RayleighData.hh]: Added more description.

**Comment by whokion** [comment]
@sethrj I updated the branch taking into account your comments.  Thank you for your second review.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighInteractor.i.hh:92]: Use `ipow<2>(cost)` for the inline "squared". And is `cost` ever really going to be less than -1? I think some extra sanity assertions (bounds checking) in here would be good.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighInteractor.i.hh:123]: Same comment as above about `exp(-n log(v))` and hard-coded doubles.

File: src/physics/em/detail/RayleighInteractor.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighInteractor.hh:78]: I like this a lot!

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [inline src/physics/em/detail/RayleighInteractor.i.hh:113]:
```suggestion
    input.factor = ipow<2>(energy * RayleighInteractor::hc_factor());
```

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:158]: Can you extend this test function so it's over a handful of energy values and test `rng.count()` to test how the rejection rate changes over the energy range of the model?

**Review by sethrj — CHANGES_REQUESTED** [review]
Looking good, Soon -- some comments about the bowels of the sampling algorithm, and  suggestions for simplifying the input data. Thanks!

File: src/physics/em/detail/Rayleigh.cc
Comment (whokion) [reply src/physics/em/detail/Rayleigh.cc:33]: Removed detail and changed rayleigh_num_parameters to RayleighData::num_parameters - Originally angular_parameter was in the global namespace inside RayleighData.hh, but I introduced RayleighData POD as you wanted to put actual parameters data in cc.  I will keep RayleighData unless you want to move data back to the header.

File: src/physics/em/detail/RayleighInteractor.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.hh:70]: I also tried to dig out where this 0.02 came from, but have failed so far - probably somewhere in a fortran code of EGCS, NPMC or whatever where the table and the sampling code were copied from (by Vladimir I.).  Based on the code, this is an arbitrary point where the functional form of the weigh in the fit change (i.e, two different fit functions).  As the unit is already taken into account in factor (unit-less as hc and energy were in the same MevEnergy).  I will add something for this limit and its unit based on my best guess now (and may try to dig out the original source farther).

File: src/physics/em/detail/Rayleigh.hh
Comment (whokion) [reply src/physics/em/detail/Rayleigh.hh:61]: Good catch and suggestion! I cannot believe myself that RayleighElementData data was not used here after it was introduced.

File: src/physics/em/detail/Rayleigh.hh
Comment (whokion) [reply src/physics/em/detail/Rayleigh.hh:37]: Added RayleighElementData 
<pre>
//---------------------------------------------------------------------------//
/*!                                                                             
 * Rayleigh angular parameters to fit tabulated form factors (\em FF)           
 * \f[                                                                          
 *  FF(E,cos)^2 = \Sigma_{j} \frac{a_j}{[1 + b_j x]^{n}}                        
 * \f]                                                                          
 * where \f$ x= E^{2}(1-cos\theta) \f$ and \em n is the high energy slope of    
 * the form factor and \em a and \em b are free parameters to obtain the best   
 * fit to the form factor. The unit for the energy (\em E) is in MeV.           
 */
struct RayleighElementData
</pre>

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh:58]: Done accordingly.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh:87]: Done and droped decimals in elsewhere - did not known that using 1.0 involves an unnecessary double conversion (actually I thought the opposite way) !

File: src/physics/em/detail/Rayleigh.cu
Comment (whokion) [reply src/physics/em/detail/Rayleigh.cu:44]: Done.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh:71]: To me too.  
- x was reused later, but agrees that they are different variables.  Used u here and use x later as need.
- unsigned int was my typo and the original code looked like this,
  https://geant4.kek.jp/lxr/source/processes/electromagnetic/lowenergy/src/G4RayleighAngularGenerator.cc#L113
- used the generate_canonical
- added a comment

File: src/physics/em/detail/RayleighInteractor.hh
Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.hh:70]: Best guess would be great! Even your statement
> an arbitrary point where the functional form of the weight in the fit change
would be helpful :)

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh:92]: Used ipow<2>(cost).  Regarding to cost, it seems that there are cases that cost < -1.0 during the sampling (so cannot asset it during the sampling).  I will leave this as it is for now.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh:123]: Done.

File: src/physics/em/detail/Rayleigh.hh
Comment (sethrj) [reply src/physics/em/detail/Rayleigh.hh:86]: I think the key is having some phrase that says "this is templated on `<Ownserhip, Memspace>` and has operator= and bool" -- which generally implies having collections, though of course doesn't have to. Something that's safe

Because the earlier Interactor work didn't have `Params` in their name we didn't have it here, but all the interactor data is effectively `Params` -- though I think for Livermore we're also passing in a shared stack allocator 🤔

Comment (whokion) [reply src/physics/em/detail/Rayleigh.hh:86]: In this case, the content (of RayleighPointer) is constant variables (model_id and gamma_id) and constant parameters (collection).  For now, I will use Group and rename it accordingly once we adopt something better later as other physics interactors will follow similar naming conventions.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.i.hh:113]: As I see it, it's a matter of telling someone "where do I look to find this function?":
- For local or  in-the-same-file functions, no namespace is needed
- For a free function in a different header, generally use `celeritas::` or `detail::` as appropriate
- For member functions of the class, use `this->`
- For static functions of the class, use `ClassName::` (though I'm OK with aliasing `This = RayleighInteractor` so you can call `This::`)

I don't want to make a big deal about it so if you think readability is *worse* by adding `RayleighInteractor::` we can skip it.

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.i.hh:87]: Yeah, it's really tricky. For a real_type of `v`:
- `v + 1 / 3` will be `v + 1` because of order-of operations (really bad!)
- `v + 0.3` will promote `v` to double-precision and do the addition in double-precision, then demote `v` back to float (bad!)
- `v + 0.5` shockingly on some compilers (GCC10 host) seems to be optimized so that it does the addition directly as a floating-point value. Same for `+ 1.0` -- I guess because 1 and .5 are exactly the same in float/double and can be converted back and forth and guaranteed to be correct.
- `v + 1` promotes the integer to a double so is fine
- `v + real_type(0.3)` will truncate 0.3 to 0.3f at compile time as needed and do the addition in single-precision.

So it looks like you're actually right in the particular case of 1.0 🤔 The more you know...

Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.i.hh:87]: And even worse `std::log(1. - f)` promotes `f` to double (then does double-precision log) and doesn't use the 1.0 shortcut! So super confusing. Easier to stick with `int` when the order-of-op doesn't get you in trouble, or add `real_type` around the literals when it might 😄

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc:183]: Extended test with multiple energy points and added rng.count().

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh]: This is more complicated than I though.  Do we need something like one<T>(), half<T>() and so on for commonly used hard coded numbers?

File: src/physics/em/detail/RayleighInteractor.i.hh
Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.i.hh]: Aha! RayleighInteractor:: is used instead of this-> since it is a static member function.  Clear enough!

**Comment by whokion** [comment]
@sethrj I tried to follow up all your comments (hopefully did not miss anything, but not sure whether I introduced other wrong stuffs).  Anyway, thanks a lot for all your time for reviewing this MR and valuable comments!

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:127]: I think this line is a null-op, and at the beginning/end of the loop you want:
```
        rng_engine.reset_count();
```

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:140]: Holy hell, a single interaction take 17K random numbers (at the low end of the energy range)?  It looks like at higher energies it's more reasonable, but could this indicate a bug in the rejection sampling? Also since the model is valid down to 1e-5 MeV (our test geant data has cross sections down to 1e-4) it might be worth testing an even lower energy to see if this explodes.

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:154]: Since this is an average I'd make it a `double`.

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:182]: Needs a `rng_engine.reset_count();`

**Review by sethrj — COMMENTED** [review]
I've got final small simplification to suggest, but also noticed that the RNG count rate blows up at low energies and I'm not sure if this is expected or not.

File: src/physics/em/detail/Rayleigh.hh
Comment (whokion) [reply src/physics/em/detail/Rayleigh.hh:92]: Sorry that I did not catch that.  I removed the RayleighParameters struct and use RayleighParameters instead of RayleighElementData (just for naming).

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc:127]: Okay, I thought that instantiating rng inside loop automatically reset the counter (but actually not).  I move the rng outside the loop and put reset_count at the beginning of the loop - I think that resetting count at the end is redundant for this case.

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc:140]: After putting reset_count, now average rng counts are
<pre>
{17774ul, 1568ul, 74ul, 8ul, 14ul};
{49710ul, 11914ul, 137ul, 12ul, 10ul};
</pre>
Agrees that there is something wrong in the low energy reason, which I was planning to investigate today and validate results with respect to Geant4 today.  As you noticed, I did not try at low energy as the test was hanging very long time, apparently stuck in the rng generation from debugging (but could be false information).  Anyway, I will spend sometime for validation.

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc:154]: Changed to real_type.

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc:182]: Added.

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [reply test/physics/em/Rayleigh.test.cc:140]: Hah I believe that following the trend you'd see an apparent hang at lower energies :) OK, let me know if you want to keep debugging on this branch, or merge it in and do validation in a separate branch. If the former I can take a closer look at the sampling implementation and compare against the reference documentation and the geant implementation.

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [reply test/physics/em/Rayleigh.test.cc:131]: That's a good idea... that would make this much simpler (in a couple of other tests as well). Let me add this to the to-do Q3 review list, and we'll implement after this gets merged.

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc]: @sethrj I will validate results today with respect to Geant4 and may understand the low sampling efficiency.  Since it smells like a bug (or involve a wrong unit conversion), please hold the merge.  Of course, you are welcome to take a closer look at the geant4 code, but busy for other stuffs.  If I need a help, I will let you know.

File: src/physics/em/detail/RayleighInteractor.hh
Comment (amandalund) [inline src/physics/em/detail/RayleighInteractor.hh:69]: I think maybe this should be `unit_cast(units::MevEnergy{1.0}`?

Comment (whokion) [reply src/physics/em/detail/RayleighInteractor.hh:69]: @amandalund You are absolutely right.  Nice catch and you save may day!

Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.hh]: Ahhh good catch indeed! I missed it because I was thinking of the recent mag field stuff :)

@whokion  you can also write this as
```
return units::centimeter * units::Mev::value();
```

**Comment by whokion** [comment]
@sethrj After a bug that was caught by @amandalund is fixed, test results look much more reasonable.  I am working on a quick validation with respect to Geant4 anyway, but you may merge this MR anytime before you leave.  If I find more issues, I will create another branch.  Thanks again for your reviews for this MR!

**Comment by sethrj** [comment]
Woohoo, great teamwork @amandalund and @whokion ! Those rejection rates look waaaaay better 😄

File: test/physics/em/Rayleigh.test.cc
Comment (sethrj) [inline test/physics/em/Rayleigh.test.cc:191]: Whoops, the fact that the averages below are integers clued me into the fact that we're being bitten by integer division in that divide-by term. Sorry for not catching that earlier!
```suggestion
        average_rng_counts.push_back(real_type(rng_engine.count()) / real_type(num_samples));
```

**Review by sethrj — APPROVED** [review]
One final small error in the test this looks good! Scattering angles as well as RNG look much more reasonable --forward-scatter at high energy, isotropic at low E :)

File: test/physics/em/Rayleigh.test.cc
Comment (whokion) [reply test/physics/em/Rayleigh.test.cc]: No problem.  Thanks for catching this.  I updated the test code accordingly.

**Comment by whokion** [comment]
@sethrj Tables included below are comparison of the scattered angle (in terms of cos(theta) from RayleighInteractor with respect to Geant4 (10.7).  I also checked intermediate variables between too codes which agrees very well.  Of course, we may further optimize code further, but the current branch should be good enough for the first integration.  Thanks again for all your review for this MR.

<pre>
<cos(theta)> (1M samples)
#-----------------------------------------------------------------
# O (Z=8)   Energy          Geant4                 celeritas
#-----------------------------------------------------------------
             1e-05     -0.000661751948424    -0.00100469639543812 
            0.0001      0.0013506838683       0.000374968514768567 
             0.001      0.0258755377214       0.0277404327481332  
              0.01      0.581895263628        0.58242081722658  
               0.1      0.951156924208        0.951278922116801  
                 1      0.999416814982        0.999413247055398  
                10      0.999994139383        0.999994105819036  
               100      0.999999941295        0.999999941220734  
              1000      0.999999999415        0.999999999419088  
#-----------------------------------------------------------------
# W (Z=74)  Energy         Geant4                 celeritas
#-----------------------------------------------------------------
             1e-05     -0.000668270231415    -0.00100161453511301 
            0.0001      0.00123504346868      0.000244991223152257
             0.001      0.0139271267733       0.0154743395737408  
              0.01      0.30900079497         0.309447347164265   
               0.1      0.819226606494        0.819057196707145   
                 1      0.982060407913        0.98198116648345    
                10      0.998711295985        0.998654084809266   
               100      0.999895820165        0.999902701458381   
              1000      0.999994680821        0.999994255633645   
#-----------------------------------------------------------------
# Pb (Z=82) Energy        Geant4                 celeritas
#-----------------------------------------------------------------
             1e-05     -0.000667710606251    -0.000998496862113119 
            0.0001      0.00123994912807      0.000296704046502483 
             0.001      0.014614573039        0.0162040738778125   
              0.01      0.294367505683        0.293245570378171    
               0.1      0.815690752791        0.815947404175699    
                 1      0.976944020836        0.976794368971151    
                10      0.997600713846        0.997570766990248    
               100      0.999741153863        0.999734462794978    
              1000      0.999975463476        0.999972650629776    
#-----------------------------------------------------------------
 </pre>

**Review by sethrj — APPROVED** [review]
Great job and thanks for the extra validation!

**Pull request merged**
